component/lambda/functions/workspace_delegations_population.py

```diff
@@ -1,336 +1,212 @@
+        insert_workspace_owner_subscriptions_results = []
 
-# Function to query owner workspaces
-def query_owner_subscriptions(lago_api_token, workspace_owner_id):
-    url = f"{lago_api_url}/v1/subscriptions?external_customer_id={workspace_owner_id}&status[]=pending&status[]=active&status[]=terminated&status[]=active"
-    response = requests.get(url, headers={"Authorization": f"Bearer {lago_api_token}"})
-    if response.status_code == 200:
-        return response.json().get('subscriptions', [])
-    else:
-        raise Exception(f"Failed to query subscriptions for workspace_owner_id {workspace_owner_id}: {response.status_code} - {response.text}")
-
-# Function to execute a SQL query and return the results
-def execute_query(query):
-    try:
-        data_response = client_redshift.execute_statement(
-            WorkgroupName='platform-app-datastore',
-            Database='data',
-            SecretArn=secret_arn,
-            Sql=query
-        )
-        data_statement_id = data_response['Id']
-
-        # Wait for the data query to complete using the helper function
-        wait_for_completion(data_statement_id, client_redshift)
+        # Get the current timestamp for record insertion
+        current_timestamp = time.strftime('%Y-%m-%d %H:%M:%S')
 
-        # Determine if the query is a SELECT or an INSERT
-        if query.strip().upper().startswith("SELECT"):
-            # Retrieve the result data for SELECT queries
-            data_result = client_redshift.get_statement_result(Id=data_statement_id)
-            # Transform the records into the desired format
-            output = transform_records(data_result['ColumnMetadata'], data_result['Records'])
-            return {
-                'statusCode': 200,
-                'body': json.dumps(output)  # Return the formatted data
-            }
-        else:
-            # For INSERT or other queries, return a success message
-            return {
-                'statusCode': 200,
-                'body': json.dumps({"message": "Query executed successfully."})
-            }
+        # Query all owner subscriptions and compare to Lago
+        latest_owner_subscriptions = cast(Iterator['LatestOwnerSubscription'], self.redshift.query("""
+            SELECT owner_pk, subscription_id, subscription_start_date, subscription_end_date, plan_code, external_id
+              FROM workspace_operations.owners
+              LEFT OUTER JOIN workspace_operations.latest_owner_subscriptions USING (owner_pk)
+        """))
+
+        for owner_pk, si_subscriptions in groupby(latest_owner_subscriptions, lambda sub: sub['owner_pk']):
+            si_subscriptions = { sub['external_id']: sub for sub in si_subscriptions }
+
+            # Get Lago subscriptions for the owner
+            lago_subscriptions = cast(LagoSubscriptionsResponse, self.lago.get(f"/api/v1/subscriptions?external_customer_id={owner_pk}&status[]=pending&status[]=active&status[]=terminated&status[]=active").json())
```

Comment:
I don't think this is a good idea to run every few minutes in the workspace delegations, I reckon we'll get rate limited by Lago or the lambda will timeout.

e.g. If we have 25000 customer records in Lago this one lambda will take at least 20 minutes to just iterate through this loop (assuming 50ms per call), where the maximum single lambda runtime is 15 minutes.

Reply:
We discussed this, and I agree :) We have a little time, but I need to wait until after the next AI turn. What we planned:

1. Separate missing workspace upload vs. subscription sync into separate lambdas
2. Update record_timestamps if nothing has changed for a customer; only pick the N most out of date customers on each run